optuna/integration/botorch.py

```diff
@@ -0,0 +1,550 @@
+        params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
+        if self._constraints_func is not None:
+            n_constraints = len(next(iter(self._trial_constraints.values())))
+            con = numpy.empty((n_trials, n_constraints), dtype=numpy.float64)
+        else:
+            con = None
+        bounds = trans.bounds
+
+        for trial_idx, trial in enumerate(trials):
+            params[trial_idx] = trans.transform(trial.params)
+            assert len(study.directions) == len(trial.values)
+
+            for obj_idx, (direction, value) in enumerate(zip(study.directions, trial.values)):
+                assert value is not None
+                if direction == StudyDirection.MINIMIZE:  # BoTorch always assumes maximization.
+                    value *= -1
+                values[trial_idx, obj_idx] = value
+
+            if con is not None:
+                con[trial_idx] = self._trial_constraints[trial_idx]
+
+        values = torch.from_numpy(values)
+        params = torch.from_numpy(params)
+        if con is not None:
+            con = torch.from_numpy(con)
+        bounds = torch.from_numpy(bounds)
+
+        if con is not None:
+            if con.dim() == 1:
+                con.unsqueeze_(-1)
+        bounds.transpose_(0, 1)
+
+        if self._candidates_func is None:
+            self._candidates_func = _get_default_candidates_func(n_objectives=study.n_objectives)
+        candidates = self._candidates_func(params, values, con, bounds)
+
+        if not isinstance(candidates, torch.Tensor):
+            raise TypeError("Candidates must be a torch.Tensor.")
+        if candidates.dim() == 2:
+            if candidates.size(0) != 1:
+                raise ValueError(
+                    "Candidates batch optimization is not supported and the first dimension must "
+                    "have size 1 if candidates is a two-dimensional tensor. Actual: "
+                    f"{candidates.size()}."
+                )
+            # Batch size is one. Get rid of the batch dimension.
+            candidates = candidates.squeeze(0)
+        if candidates.dim() != 1:
+            raise ValueError("Candidates must be one or two-dimensional.")
+        if candidates.size(0) != bounds.size(1):
+            raise ValueError(
+                "Candidates size must match with the given bounds. Actual candidates: "
+                f"{candidates.size(0)}, bounds: {bounds.size(1)}."
+            )
+
+        candidates = candidates.numpy()
+
+        params = trans.untransform(candidates)
+
+        # Exclude upper bounds for parameters that should have their upper bounds excluded.
+        # TODO(hvy): Remove this exclusion logic when it is handled by the data transformer.
+        for name, param in params.items():
+            if isinstance(search_space[name], (UniformDistribution, LogUniformDistribution)):
+                params[name] = min(params[name], search_space[name].high - 1e-8)
```

Comment:
`1e-8` may be too large when users use `LogUniformDistribution` since users can specify `1e-8` as the value of `high`. How about shifting the value in the log-space, for example, as follows?

```suggestion
            if isinstance(search_space[name], UniformDistribution):
                params[name] = min(params[name], search_space[name].high - 1e-8)
            elif isinstance(search_space[name], LogUniformDistribution):
                shifted_high = math.exp(math.log(search_space[name].high) - 1e-8)
                params[name] = min(params[name], shifted_high)
```

Reply:
Fixed.

Reply:
When users specify these log distributions, are you doing any transformations of the inputs so that the models are fit on a log scale-inputs?  That would be really helpful for achieving good model performance.  One advantage of Ax is that a lot of the bookkeeping for these transforms are handled automatically (you just have to specify log_scale: True when defining your search space).

Reply:
Yes, the data and the bounds are transformed accordingly before being passed to the functions that construct the models. So it should behave as you describe. There is a [helpers in Optuna that does the transformation](https://github.com/optuna/optuna/blob/master/optuna/_transform.py) such as one-hot encodings and discretizations. We could perhaps rely on Ax to do this but it's coupled to the data structures (distribution classes) in Optuna and it's a lower level utility to be used by several samplers. It might be good to be aware of inconsistencies if there are such.

(In Optuna, a users specifies this log distributions, discretizations, etc. in the [suggest methods](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.trial.Trial.html#optuna.trial.Trial.suggest_float))